### PR TITLE
Refactor scoring PoWs and deprecate old hash functions

### PR DESCRIPTION
**Pull request by grtlr**
# Description of change

This is the first step of bringing `bee-protocol` over to the `dev` branch. By storing the hash functions in `ProofOfWork`, we can call `ProofOfWork::hash` multiple times without performance hits. The next step would be to use `bee-pow` in the `chrysalis-pt2` branch.

A nice benefit would be that the hashing rate would be easier to benchmark, because currently we need to rebuild the complete node if we change small pieces in the `Hasher`.

## Links to any relevant issues

This PR is a first step towards #829.

## Type of change

- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Ran all test cases.

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

**Review by pvdrz — DISMISSED** [review]

**Comment by grtlr** [comment]
If we merge this, it might make sense to push that version to `crates.io` so that potential users see the deprecation of these methods.

File: bee-pow/Cargo.toml
Comment (pvdrz) [inline bee-pow/Cargo.toml]: Does this block the PR?

**Review by pvdrz — DISMISSED** [review]

File: bee-pow/Cargo.toml
Comment (grtlr) [reply bee-pow/Cargo.toml]: Now that @thibault-martinez has bumped the version of `iota-crypto` 💪 everything is fine.

File: bee-crypto/Cargo.toml
Comment (thibault-martinez) [inline bee-crypto/Cargo.toml]:
```suggestion
iota-crypto = { version = "0.8.0", default-features = false, features = [ "blake2b", "digest", "curl-p" ] }
```

Comment (grtlr) [reply bee-crypto/Cargo.toml]: Oops :)

**Comment by grtlr** [comment]
We should prepare the corresponding PR on `chrysalis-pt2` before we merge this, otherwise our CI might start failing there.

File: bee-signing/tests/wots_sponge.rs
Comment (thibault-martinez) [inline bee-signing/tests/wots_sponge.rs]: Merge both imports

File: bee-signing/tests/ternary_seed.rs
Comment (thibault-martinez) [inline bee-signing/tests/ternary_seed.rs]: I think I would even deprecate Sponge tbh... The Curl version in iota-crypto doesn't implement this trait.

File: bee-pow/Cargo.toml
Comment (thibault-martinez) [inline bee-pow/Cargo.toml]:
```suggestion
iota-crypto = { version = "0.8.0", default-features = false, features = [ "blake2b", "digest", "curl-p" ] }
```

File: bee-crypto/tests/sponge.rs
Comment (thibault-martinez) [inline bee-crypto/tests/sponge.rs]: If we deprecate all current sponge kinds, there is not reason not to deprecate `SpongeKind` ?

File: bee-crypto/CHANGELOG.md
Comment (thibault-martinez) [inline bee-crypto/CHANGELOG.md]: Can you actually prepare a release ? Maybe `0.3.0`, it's probably time to drop the alpha tag anyway.

File: bee-pow/CHANGELOG.md
Comment (thibault-martinez) [inline bee-pow/CHANGELOG.md]: Same, prepare a release, 0.2.0

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Comment by grtlr** [comment]
> We should prepare the corresponding PR on `chrysalis-pt2` before we merge this, otherwise our CI might start failing there.

I've checked, this PR _shouldn't_ cause any problems with `chrysalis-pt2`.

File: Cargo.toml
Comment (grtlr) [inline Cargo.toml:33]: Maybe this can be changed again back when we start handling `Cargo.lock` files differently.

File: bee-message/Cargo.toml
Comment (grtlr) [inline bee-message/Cargo.toml]: Wait for #844 and https://github.com/iotaledger/crypto.rs/pull/118

```suggestion
iota-crypto = { version = "0.9.1", default-features = false, features = [ "ed25519", "blake2b" ] }
```

File: bee-signing/CHANGELOG.md
Comment (thibault-martinez) [inline bee-signing/CHANGELOG.md]: Change date (hopefully today, should be fine)

File: bee-pow/CHANGELOG.md
Comment (thibault-martinez) [inline bee-pow/CHANGELOG.md]: Same

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: Cargo.lock
Comment (thibault-martinez) [inline Cargo.lock]: Any idea why this is not 0.9.1 ?

File: bee-crypto/Cargo.toml
Comment (thibault-martinez) [inline bee-crypto/Cargo.toml]: Alphanumeric order please :)

File: bee-crypto/CHANGELOG.md
Comment (thibault-martinez) [inline bee-crypto/CHANGELOG.md]: Same

**Review by thibault-martinez — DISMISSED** [review]

**Review by pvdrz — DISMISSED** [review]

File: bee-pow/src/score.rs
Comment (Alex6323) [inline bee-pow/src/score.rs]: I think you could also use the `append` API here on `TritBuf`:

```
self.pow_input.append(b1t6::encode::<T1B1buf>(&pow_digest).as_slice());
self.pow_input.append(b1t6::encode::<T1B1buf>(tail).as_slice());
```
Maybe to reduce repetition also create a slice of the 3 `Btrit::Zero` and append it as well. :slightly_smiling_face:

**Review by Alex6323 — DISMISSED** [review]

**Review by Alex6323 — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**
